app/shared/forever-ui/src/main/kotlin/com/hedvig/android/shared/foreverui/ui/ui/ForeverDestination.kt

```diff
@@ -114,159 +143,177 @@ private fun ForeverScreen(
+}
+
+@Composable
+internal fun LoadingForeverContent() {
+  Column(
+    Modifier
+      .fillMaxSize()
+      .verticalScroll(rememberScrollState())
+      .windowInsetsPadding(WindowInsets.safeDrawing.only(WindowInsetsSides.Horizontal)),
+  ) {
+    Spacer(Modifier.windowInsetsTopHeight(WindowInsets.safeDrawing))
+    Row(
+      horizontalArrangement = Arrangement.SpaceBetween,
+      verticalAlignment = Alignment.CenterVertically,
+      modifier = Modifier
+        .height(64.dp)
+        .fillMaxWidth()
+        .padding(horizontal = 16.dp),
+    ) {
+      HedvigText(
+        text = stringResource(R.string.TAB_REFERRALS_TITLE),
+        style = HedvigTheme.typography.headlineSmall,
+      )
+    }
+    Spacer(Modifier.height(16.dp))
+    HedvigText(
+      text = "-",
+      textAlign = TextAlign.Center,
+      color = HedvigTheme.colorScheme.textSecondary,
+      modifier = Modifier
+        .withoutPlacement()
+        .fillMaxWidth(),
+    )
+    Spacer(Modifier.height(16.dp))
+    DiscountPieChart(
+      totalPrice = 0f,
+      totalExistingDiscount = 0f,
+      incentive = 0f,
+      modifier = Modifier
+        .padding(horizontal = 16.dp)
+        .fillMaxWidth()
+        .wrapContentWidth(Alignment.CenterHorizontally)
+        .clip(CircleShape)
+        .size(215.dp)
+        .placeholder(
+          visible = true,
+          highlight = PlaceholderHighlight.shimmer(),
+        ),
     )
+    Spacer(Modifier.weight(1f))
   }
 }
 
-@ExperimentalMaterial3Api
 @Composable
 internal fun ForeverContent(
-  uiState: ForeverUiState,
+  uiState: ForeverUiState.Success,
   pullRefreshState: PullRefreshState,
   onShareCodeClick: (code: String, incentive: UiMoney) -> Unit,
   onSubmitCode: (String) -> Unit,
   showedReferralCodeSubmissionError: () -> Unit,
-  showedCampaignCodeSuccessfulChangeMessage: () -> Unit,
+  showedReferralCodeSuccessfulChangeMessage: () -> Unit,
+  openEditCodeBottomSheet: () -> Unit,
+  closeEditCodeBottomSheet: () -> Unit,
 ) {
-  val coroutineScope = rememberCoroutineScope()
-  var textFieldValueState by remember(uiState.foreverData?.campaignCode) {
-    mutableStateOf(TextFieldValue(text = uiState.foreverData?.campaignCode ?: ""))
+  val initialTextValue = TextFieldValue(
+    text = uiState.foreverData?.campaignCode ?: "",
+    selection = TextRange((uiState.foreverData?.campaignCode ?: "").length),
+  )
+  var textFieldValue by remember(uiState.foreverData?.campaignCode) {
+    mutableStateOf(initialTextValue)
   }
 
-  LaunchedEffect(Unit) {
-    snapshotFlow { textFieldValueState }.collectLatest {
-      showedReferralCodeSubmissionError() // clear error after the member edits the code manually
+  LaunchedEffect(textFieldValue) {
+    showedReferralCodeSubmissionError() // Clear error on new referral code input
+  }
+  LaunchedEffect(uiState.showEditReferralCodeBottomSheet) {
+    if (!uiState.showEditReferralCodeBottomSheet) {
+      delay(500) // to avoid extra animation effects on closing bottom sheet
```

Comment:
// todo: okay, so this delay is ugly, I admit, but it's the only way to restore the value back on closing the BS and don't have a weird animation effect when closing the bottom sheet
while having submit error/changed textFieldValue in the BS